lib/src/gitattributes.rs

```diff
@@ -0,0 +1,705 @@
+                message: message.clone(),
+                source: message.into(),
+            }),
+        }
+    }
+}
+#[cfg(test)]
+mod tests {
+
+    use gix_attributes::state::Value;
+
+    use super::*;
+
+    fn repo_path(path: &str) -> &RepoPath {
+        RepoPath::from_internal_string(path).unwrap()
+    }
+
+    fn create_git_attributes(files: &[(&'static str, &'static str)]) -> GitAttributes {
+        let data: TestStore = files
+            .iter()
+            .map(|(key, value)| {
+                (
+                    RepoPathBuf::from_internal_string(key.to_string()).unwrap(),
+                    Ok(value.to_string()),
+                )
+            })
+            .collect();
+
+        GitAttributes::new(HashMap::new(), data)
+    }
+
+    // macros so tests fail on the right line
+    macro_rules! assert_search_output {
+        ($git_attributes: expr,
+            $file: expr,
+            $attribute: expr,
+            $expected: expr
+        ) => {
+            assert_search_output!(
+                $git_attributes,
+                $file,
+                $attribute,
+                $expected,
+                SearchPriority::Disk
+            );
+        };
+        ($git_attributes: expr,
+            $file: expr,
+            $attribute: expr,
+            $expected: expr,
+            $search: expr
+        ) => {
+            let map = $git_attributes
+                .search(repo_path($file), &[$attribute], $search)
+                .block_on()
+                .unwrap();
+            assert_eq!(map.len(), 1);
+            assert_eq!(*map.get($attribute).unwrap(), $expected);
+        };
+    }
+    // macros so tests fail on the right line
+    macro_rules! assert_attribute_output {
+        ($git_attributes_content: expr,
+            $file: expr,
+            $attribute: expr,
+            $expected: expr
+        ) => {
+            let git_attributes =
+                create_git_attributes(&[(".gitattributes", $git_attributes_content)]);
+            assert_search_output!(&git_attributes, $file, $attribute, $expected)
+        };
+    }
+
+    #[test]
+    fn test_search_state_set() {
+        assert_attribute_output!("abc foo", "abc", "foo", State::Set);
+    }
+
+    #[test]
+    fn test_search_state_eol_lf() {
+        assert_attribute_output!("foo eol=lf", "foo", "eol", State::Value(Value::from("lf")));
+    }
+
+    #[test]
+    fn test_search_unset() {
+        assert_attribute_output!("foo -text", "foo", "text", State::Unset);
+    }
+
+    #[test]
+    fn test_search_unspecified() {
+        assert_attribute_output!("!foo -text", "foo", "text", State::Unspecified);
+    }
+
+    #[test]
+    fn test_search_unspecified_no_pattern() {
+        assert_attribute_output!("foo elo=lf", "bar", "text", State::Unspecified);
+        assert_attribute_output!("foo elo=lf", "bar", "elo", State::Unspecified);
+    }
+
+    #[test]
+    fn path_and_pattern_matching() {
```

Comment:
According to my sampling from `jj`, almost all the tests are prefixed with `test_`. Can we rename the test function with `test_` prefixed?